contracts/tvl-staking/RibbonDualStakingRewards.sol

```diff
@@ -0,0 +1,359 @@
+    _balances[msg.sender] = _balances[msg.sender].sub(amount);
+    if (block.timestamp < periodFinish.add(1 days)) {
+      delete rewards[msg.sender];
+    }
+    stakingToken.safeTransfer(msg.sender, amount);
+    emit Withdrawn(msg.sender, amount);
+  }
+
+  function getReward() public override nonReentrant updateReward(msg.sender) {
+    if (block.timestamp >= periodFinish.add(1 days)) {
+      Rewards memory _rewards = rewards[msg.sender];
+      if (_rewards.token0 > 0) {
+        rewardsToken0.safeTransfer(
+          msg.sender,
+          Math.min(_rewards.token0, rewardsToken0.balanceOf(address(this)))
+        );
+        emit RewardPaid(address(rewardsToken0), msg.sender, _rewards.token0);
+      }
+      if (_rewards.token1 > 0) {
+        rewardsToken1.safeTransfer(
+          msg.sender,
+          Math.min(_rewards.token1, rewardsToken1.balanceOf(address(this)))
+        );
+        emit RewardPaid(address(rewardsToken1), msg.sender, _rewards.token1);
+      }
+      delete rewards[msg.sender];
+    }
+  }
+
+  function exit() external override {
+    withdraw(_balances[msg.sender]);
+    getReward();
+  }
+
+  function _numWeeksPassed(uint256 time) internal view returns (uint256) {
+    uint256 _startEmission = startEmission;
+    if (time < _startEmission) {
+      return 0;
+    }
+    return time.sub(_startEmission).div(1 weeks).add(1);
+  }
+
+  /* ========== RESTRICTED FUNCTIONS ========== */
+
+  function notifyRewardAmount(uint256 reward0, uint256 reward1)
+    external
+    override
+    onlyRewardsDistribution
+    updateReward(address(0))
+  {
+    Rewards memory _rewardRate;
+    uint256 _rewardsDuration = rewardsDuration;
+    if (block.timestamp >= periodFinish) {
+      _rewardRate = Rewards(
+        reward0.div(_rewardsDuration).toUint128(),
+        reward1.div(_rewardsDuration).toUint128()
+      );
+    } else {
+      uint256 remaining = periodFinish.sub(block.timestamp);
+      _rewardRate = rewardRate;
+      _rewardRate = Rewards(
+        reward0
+          .add(remaining.mul(_rewardRate.token0))
+          .div(_rewardsDuration)
+          .toUint128(),
+        reward1
+          .add(remaining.mul(_rewardRate.token1))
+          .div(_rewardsDuration)
+          .toUint128()
+      );
+    }
+
+    // Ensure the provided reward amount is not more than the balance in the contract.
+    // This keeps the reward rate in the right range, preventing overflows due to
+    // very high values of rewardRate in the earned and rewardsPerToken functions;
+    // Reward + leftover must be less than 2^256 / 10^18 to avoid overflow.
+    require(
+      _rewardRate.token0 <=
+        rewardsToken0.balanceOf(address(this)).div(rewardsDuration),
+      "Provided reward0 too high"
+    );
+    require(
+      _rewardRate.token1 <=
+        rewardsToken1.balanceOf(address(this)).div(rewardsDuration),
+      "Provided reward1 too high"
+    );
+
+    rewardRate = _rewardRate;
```

Comment:
you cant set a storage ref to a memory. have to do like this:

```
rewardRate.token0 = _rewardRate .token0;
rewardRate.token1 = _rewardRate .token1;
```

Reply:
Wdym? `_rewardRate` is initialized in memory and then mutated & set to storage
```
rewardRate = _rewardRate;
```
does the same thing as
```
rewardRate.token0 = _rewardRate.token0;
rewardRate.token1 = _rewardRate.token1;
```
(I just checked it now using console.log)

Also it seems 
```
_rewardRate.token0 = reward0.div(_rewardsDuration).toUint128();
_rewardRate.token1 = reward1.div(_rewardsDuration).toUint128();
```
uses less gas than
```
_rewardRate = Rewards(
   reward0.div(_rewardsDuration).toUint128(),
   reward1.div(_rewardsDuration).toUint128()
);
```
so i've changed it here: https://github.com/ribbon-finance/governance/pull/46/commits/e6a99a94734d3db562a46d6ff0d74d3e02361711

Reply:
Oh if it works it works. I was in a similar spot with one of the governance contracts where I was setting a storage struct to a memory struct and changes were not reflected. so I had to assign variable by variable.